### PR TITLE
Enhancement/content api

### PR DESCRIPTION
**Pull request by Figments**
The first of many relating to the migration to the new content API.

This one covers blogs.

### CODE REVIEW

File: packages/client/src/app/pages/portfolio/my-stuff/blog-form/blog-form.component.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  submitForm() {
+    if (this.fields.title.invalid) {
+      return;
+    }
+
+    if (this.fields.body.invalid) {
+      return;
+    }
```
Comment (pingzing) [inline packages/client/src/app/pages/portfolio/my-stuff/blog-form/blog-form.component.ts]: Should pop up errors here.

File: packages/client/src/app/pages/portfolio/my-stuff/blog-form/blog-form.component.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    const formData: BlogForm = {
+      title: this.fields.title.value,
+      body: this.fields.body.value
+    };
+
+    if (this.createBlogMode) {
+      this.blogsService.createBlog(formData).subscribe(() => {
+        this.location.back();
```
Comment (pingzing) [inline packages/client/src/app/pages/portfolio/my-stuff/blog-form/blog-form.component.ts:74]: Relative navigation strikes me as a little dangerous, but I'm not totally clear on the flow here so ehhhh

File: packages/client/src/app/pages/portfolio/my-stuff/my-stuff.component.html [hunk 3]
```diff
@@ -0,0 +1,185 @@
+<div class="tools">
+    <ng-container *ngIf="route.children.length === 0; else goBack">
+        <div class="tool-buttons">
+            <!--Creation tools-->
+            <button matRipple class="tool left" [matMenuTriggerFor]="createMenu"><i-feather name="loader"></i-feather>Create</button>
```
Comment (pingzing) [inline packages/client/src/app/pages/portfolio/my-stuff/my-stuff.component.html]: This needs a tooltip

File: packages/client/src/app/pages/portfolio/my-stuff/my-stuff.component.html [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            <div class="search-bar">
+                <input type="text" formControlName="query" class="search-input" placeholder="Search your stuff...">
+                <button matRipple class="search-submit" type="submit"><i-feather name="search"></i-feather></button>
+            </div>
+        </form>
+    </ng-container>
+    <ng-template #goBack>
+        <div class="tool-buttons">
+            <!--Creation tools-->
+            <button matRipple class="tool left" [routerLink]="['/portfolio', currentUser._id, currentUser.username | slugify, 'my-stuff']"><i-feather name="chevron-left"></i-feather>Back</button>
```
Comment (pingzing) [inline packages/client/src/app/pages/portfolio/my-stuff/my-stuff.component.html]: This needs a tooltip

File: packages/client/src/app/pages/portfolio/my-stuff/my-stuff.component.html [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        
+    <!--User has content-->
+    <ng-template #hasContent>
+        <div class="content-grid">
+            <ng-container *ngIf="myContent.length === 0; else hasContentTop">
+                <div class="empty" style="width: 100%;">
+                    <h3>It's pretty lonely here...</h3>
+                    <p>Why not add some content by hitting the create button?</p>
+                </div>
+            </ng-container>
```
Comment (pingzing) [inline packages/client/src/app/pages/portfolio/my-stuff/my-stuff.component.html]: What's the distinction between these two? It seems like they're distinct forms of UI that are triggered by the same condition.

File: packages/client/src/app/services/user/my-stuff.service.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+import { HttpClient } from '@angular/common/http';
+import { throwError } from 'rxjs';
+import { map, catchError } from 'rxjs/operators';
+import { Types } from 'mongoose';
+import { BlogsContentModel, ContentKind, ContentModel, Folder, FolderForm } from '@pulp-fiction/models/content';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class MyStuffService {
```
Comment (pingzing) [inline packages/client/src/app/services/user/my-stuff.service.ts:10]: love this name

File: packages/models/src/lib/content/content.model.ts [hunk 7]
```diff
@@ -14,10 +17,13 @@ export interface ContentModel {
         readonly comments: number;
     };
     audit: {
+        published: PubStatus;
+        publishedOn: Date;
         hasComments: boolean;
         isDeleted: boolean;
+        selected: boolean;
```
Comment (pingzing) [inline packages/models/src/lib/content/content.model.ts]: this feels like a view property that doesn't really belong on a data model

File: packages/models/src/lib/content/folders/folder.model.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import { UserInfo } from '@pulp-fiction/models/users';
+import { Types } from 'mongoose';
+
+export interface Folder {
+    readonly _id: Types.ObjectId;
```
Comment (pingzing) [inline packages/models/src/lib/content/folders/folder.model.ts]: (should we start using this for `_id`s everywhere?)

File: packages/models/src/lib/content/news/news-content.model.ts [hunk 9]
```diff
@@ -8,9 +10,10 @@ export interface NewsContentModel extends ContentModel {
     };
     audit: {
         featured: boolean;
-        published: boolean;
+        published: PubStatus;
         publishedOn: Date;
         isDeleted: boolean;
         hasComments: boolean;
+        selected: boolean;
```
Comment (pingzing) [inline packages/models/src/lib/content/news/news-content.model.ts]: ditto above comment re: view property

File: packages/server/src/app/db/content/blogs/blogs.service.ts [hunk 10]
```diff
@@ -1,4 +1,87 @@
...
+export class BlogsService {
+    constructor(@InjectModel('BlogContent') private readonly blogsModel: PaginateModel<BlogsContentDocument>,
+        private readonly usersService: UsersService) {}
+
+    /**
+     * Creates a new blogpost and saves it to the database. Returns the newly
+     * created blog as a promise.
+     *
+     * @param user The user making the blog.
+     * @param blogInfo The blog's information.
```
Comment (pingzing) [inline packages/server/src/app/db/content/blogs/blogs.service.ts:23]: Missing comment doc for `parentId`. Also, what _is_ a parent, in the context of a blog?

File: packages/server/src/app/db/content/content.service.ts [hunk 11]
```diff
@@ -96,4 +121,25 @@ export class ContentService {
...
+    /**
+     * Adds a view to some content.
+     * 
+     * @param contentId The content's ID
+     */
+    async addView(contentId: string) {
+        return await this.contentModel.updateOne({'_id': contentId}, {
+            $inc: {'stats.views': 1}
+        });
+    }
```
Comment (pingzing) [inline packages/server/src/app/db/content/content.service.ts]: This method name confused the heck out of me until I saw the implementation. I kept thinking it was adding a _UI_ view to something. Should probably rename it to `incrementViewCount()` or something.

File: packages/server/src/app/db/users/users.service.ts [hunk 12]
```diff
@@ -185,6 +186,21 @@ export class UsersService {
...
     }
 
+    /**
+     * Updates the number of blogs a user has tied to their account. Should only count blogs
+     * that are published and not deleted.
+     * 
+     * @param user The user to update
+     * @param addOne Whether to add one or subtract one
+     */
+    async changeBlogCount(user: JwtPayload, addOne: boolean): Promise<void> {
```
Comment (pingzing) [inline packages/server/src/app/db/users/users.service.ts:195]: The doc and name make me think this sets the count arbitrarily, but it's really just incrementing or decrementing. The name/commentdoc should probably be changed to reflect that.

**Review by pingzing — CHANGES_REQUESTED** [review]

**Comment by pingzing** [comment]
And some UI feedback:
image

Icons here are pretty blurry at 100% DPI, the coffee cup especially.

**Comment by pingzing** [comment]
image
Ditto the icons here. In addition, why do they react to mouseover if they're not buttons?

File: packages/client/src/app/pages/portfolio/my-stuff/blog-form/blog-form.component.ts (hunk 2 above)
Comment (Figments) [reply packages/client/src/app/pages/portfolio/my-stuff/blog-form/blog-form.component.ts:74]: relative navigation was hella messy with folders, but I think for now it's not really a big issue

File: packages/models/src/lib/content/folders/folder.model.ts (hunk 8 above)
Comment (Figments) [reply packages/models/src/lib/content/folders/folder.model.ts]: only where the documents we're creating aren't ones where we care about URL-friendly IDs

**Review by pingzing — APPROVED** [review]

**Pull request merged**